s3/src/main/java/ch/cyberduck/core/sts/STSCredentialsRequestInterceptor.java

```diff
@@ -0,0 +1,206 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+import ch.cyberduck.core.Credentials;
+import ch.cyberduck.core.DisabledCancelCallback;
+import ch.cyberduck.core.Host;
+import ch.cyberduck.core.LoginCallback;
+import ch.cyberduck.core.aws.CustomClientConfiguration;
+import ch.cyberduck.core.exception.BackgroundException;
+import ch.cyberduck.core.exception.InteroperabilityException;
+import ch.cyberduck.core.exception.LoginFailureException;
+import ch.cyberduck.core.oauth.OAuth2RequestInterceptor;
+import ch.cyberduck.core.preferences.HostPreferences;
+import ch.cyberduck.core.s3.S3Session;
+import ch.cyberduck.core.ssl.ThreadLocalHostnameDelegatingTrustManager;
+import ch.cyberduck.core.ssl.X509KeyManager;
+import ch.cyberduck.core.ssl.X509TrustManager;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpException;
+import org.apache.http.client.HttpClient;
+import org.apache.http.protocol.HttpContext;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jets3t.service.security.AWSSessionCredentials;
+
+import java.io.IOException;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AnonymousAWSCredentials;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenService;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceClientBuilder;
+import com.amazonaws.services.securitytoken.model.AWSSecurityTokenServiceException;
+import com.amazonaws.services.securitytoken.model.AssumeRoleWithWebIdentityRequest;
+import com.amazonaws.services.securitytoken.model.AssumeRoleWithWebIdentityResult;
+import com.google.api.client.auth.oauth2.Credential;
+
+public class STSCredentialsRequestInterceptor extends OAuth2RequestInterceptor {
+    private static final Logger log = LogManager.getLogger(STSCredentialsRequestInterceptor.class);
+
+    private final X509TrustManager trust;
+    private final X509KeyManager key;
+    private final LoginCallback prompt;
+    private final S3Session session;
+
+    private long stsExpiryInMilliseconds;
+
+    public STSCredentialsRequestInterceptor(HttpClient client, Host host, final X509TrustManager trust, final X509KeyManager key,
+                                            LoginCallback prompt, S3Session session) {
+        super(client, host);
+        this.trust = trust;
+        this.key = key;
+        this.prompt = prompt;
+        this.session = session;
+    }
+
+    @Override
+    public void process(final org.apache.http.HttpRequest request, final HttpContext context) throws HttpException, IOException {
+        if(System.currentTimeMillis() >= stsExpiryInMilliseconds) {
+            try {
+                if(tokens.isExpired()) {
```

Comment:
It should be possible to just call `super.process` here to refresh the OAuth tokens.

Reply:
To me this is not possible because in the super refresh there is no catch for the specific exceptions to do an oAuth authorization afterwards. Additionally, it sets a wrong AUTHORIZATION Header which is not a problem because this Header is recalculated in an interceptor later in the chain but to me it is not nice.

Reply:
I definitely see the point of avoiding code duplication. I don't see how to solve that with the `OAuth2ErrorResponseInterceptor`. This is an `ServiceUnavailableRetryStrategy` and reacts on 401 SC Response but we decided to refresh the tokens preemptively in an interceptor.
I thought about extracting the code to a separate method but there are always tiny differences that lead to incompatibility.